kaldb/src/main/java/com/slack/kaldb/server/KaldbIndexer.java

```diff
@@ -70,60 +41,114 @@ public static LogMessageTransformer getLogMessageTransformer(String dataTransfor
+   * which would result in wasted work.
    *
    * <p>On an indexer restart, we should start indexing at a last known good offset for that
-   * partition. If a last known good offset doesn't exist since we are consuming for the first time
-   * then we start with head. If the offset exists but the offset expired, we are in a whole world
-   * of pain. The best option may to start indexing at oldest. Or we can also start indexing at
-   * head.
-   *
-   * <p>Currently, we don't have a durable metadata store and the kafka consumer offset acts as a
-   * weak place holder. On an indexer shutdown it is very important that we ensure that we persisted
-   * * the offset of the data correctly. So we can pick up from the same location and start from
-   * that place.
-   *
-   * <p>The best way to close an indexer is the following steps: stop ingestion, index the ingested
-   * messages, persist the indexed messages and metadata successfully and then close the
-   * chunkManager and then the consumer,
+   * partition. The indexer pre-start job, ensures that we choose the correct start offset and end
+   * offset for the indexer. Optionally, the pre start task also creates an recovery task if needed,
+   * since the indexer may not be able to catch up.
    */
-  public KaldbIndexer(IndexingChunkManager<LogMessage> chunkManager, KaldbKafkaWriter kafkaWriter) {
+  public KaldbIndexer(
+      IndexingChunkManager<LogMessage> chunkManager,
+      MetadataStore metadataStore,
+      KaldbConfigs.IndexerConfig indexerConfig,
+      KaldbConfigs.KafkaConfig kafkaConfig,
+      MeterRegistry meterRegistry) {
     checkNotNull(chunkManager, "Chunk manager can't be null");
+    this.metadataStore = metadataStore;
+    this.indexerConfig = indexerConfig;
+    this.kafkaConfig = kafkaConfig;
+    this.meterRegistry = meterRegistry;
+
+    // Create a chunk manager
     this.chunkManager = chunkManager;
-    this.kafkaWriter = kafkaWriter;
+    // set up indexing pipelne
+    LogMessageTransformer messageTransformer =
+        DATA_TRANSFORMER_MAP.get(indexerConfig.getDataTransformer());
+    LogMessageWriterImpl logMessageWriterImpl =
+        new LogMessageWriterImpl(chunkManager, messageTransformer);
+    this.kafkaConsumer =
+        KaldbKafkaConsumer.fromConfig(kafkaConfig, logMessageWriterImpl, meterRegistry);
   }
 
   @Override
   protected void startUp() throws Exception {
-    LOG.info("Starting indexing into Kaldb.");
-    kafkaWriter.awaitRunning(DEFAULT_START_STOP_DURATION);
+    LOG.info("Starting Kaldb indexer.");
+    long startOffset = indexerPreStart();
+    chunkManager.awaitRunning(DEFAULT_START_STOP_DURATION);
+    // Set the Kafka offset and pre consumer for consumption.
+    kafkaConsumer.prepConsumerForConsumption(startOffset);
+    LOG.info("Started Kaldb indexer.");
+  }
+
+  /** Indexer pre-start Clean up stale indexer tasks and optionally create a recovery task */
```

Comment:
Nit - looks like the formatting clobbered a newline here? Might want to fix this.